### ISSUE TITLE
Handle PartitionsDefinition changes

### ISSUE THREAD

**Issue by hebo-yang** [dagster-io/dagster#19182]
### Dagster version

0.15.9

### What's the issue?

If an asset job's partition changed from HourlyPartitionsDefinition to DailyPartitionsDefinition, the Dagster UI still shows all the hourly partitions for the job. I restarted both my code server and dagit pods but it didn't help. 

However, if I do this same change from my local, hitting Reload Location from Dagit would have the partitions correctly regenerated. 

Dagit seems to call AssetNodeOrError graphql query. Are the asset nodes serialized to database? I am wondering if there is any way to handle such changes gracefully? If the user tries to backfill from the Dagit for a partition that is no longer valid, it would throw. 

### What did you expect to happen?

If a job/asset's partition definition changes, the UI should correctly display the partition sets for updated definition and prevent users to backfill partitions that are no longer valid. 

### Deployment type

Dagster Helm chart

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by sryza** [comment]
Am I understanding correctly that this works in your local deployment, but doesn't work in your helm deployment? In your helm deployment, after re-deploying your code server, did you try hitting the button to reload your definitions?

**Comment by hebo-yang** [comment]
> Am I understanding correctly that this works in your local deployment, but doesn't work in your helm deployment? In your helm deployment, after re-deploying your code server, did you try hitting the button to reload your definitions?

Right, the problem only happens on the helm deployment. It makes me suspect that the partition set data is persisted to the postgres db but I couldn't find where. I re-deployed code server, redeployed dagit, and reload definitions but still getting the old partitions showing up.

I checked from browser network trace that this assetNodeOrError graphQL query retrieved partitionKeysByDimension which contained the obsolete hourly partition data. I couldn't figure out from the Dagster code base where this function and query come from and where is the data stored. 

Query:
```
query PartitionHealthQuery($assetKey: AssetKeyInput!) {
  assetNodeOrError(assetKey: $assetKey) {
    ... on AssetNode {
      id
      partitionKeysByDimension {
        name
        type
        partitionKeys
        __typename
      }
```

Data:
```
            "partitionKeysByDimension": [
                {
                    "name": "default",
                    "type": "TIME_WINDOW",
                    "partitionKeys": [
                        "2020-05-31-17",
                        "2020-05-31-18",
                        "2020-05-31-19",
                        "2020-05-31-20",
                        "2020-05-31-21",
                        ...
```

On the other side, from the LaunchPad view, the partition selection drop down is getting the correct partitions. I checked that it uses this partitionSetOrError query instead:

```
query ConfigPartitionsQuery($repositorySelector: RepositorySelector!, $partitionSetName: String!, $assetKeys: [AssetKeyInput!]) {
  partitionSetOrError(
    repositorySelector: $repositorySelector
    partitionSetName: $partitionSetName
  ) {
```

**Comment by hebo-yang** [comment]
@sryza Yes, adding new asset or job are reflected. Changing asset partition definition reflects in some queries. I think it's just something strange with this query
```
query PartitionHealthQuery($assetKey: AssetKeyInput!) {
  assetNodeOrError(assetKey: $assetKey) {
```
